samza-core/src/main/java/org/apache/samza/startpoint/StartpointManager.java

```diff
@@ -173,71 +205,148 @@ public void deleteStartpoint(SystemStreamPartition ssp, TaskName taskName) {
+      String fanOutKey = toFanOutStoreKey(taskName);
+      StartpointFanOutPerTask newFanOut = fanOuts.get(taskName);
+      fanOutStore.put(fanOutKey, objectMapper.writeValueAsBytes(newFanOut));
+    }
+
+    for (SystemStreamPartition ssp : deleteKeys.keySet()) {
+      for (TaskName taskName : deleteKeys.get(ssp)) {
+        if (taskName != null) {
+          deleteStartpoint(ssp, taskName);
+        } else {
+          deleteStartpoint(ssp);
         }
       }
     }
 
-    // Delete SSP-only keys
-    sspsToDelete.forEach(ssp -> {
-        deleteStartpoint(ssp);
-        LOG.info("All Startpoints for SSP: {} have been grouped to the appropriate tasks and the SSP was deleted.");
-      });
+    return ImmutableSet.copyOf(fanOuts.keySet());
+  }
 
-    return ImmutableSet.copyOf(sspsToDelete);
+  /**
+   * Read the fanned out {@link Startpoint}s for the given {@link TaskName}
+   * @param taskName to read the fan out Startpoints for
+   * @return fanned out Startpoints
+   */
+  public Map<SystemStreamPartition, Startpoint> getFanOutForTask(TaskName taskName) throws IOException {
+    Preconditions.checkState(!stopped, "Underlying metadata store not available");
+    Preconditions.checkNotNull(taskName, "TaskName cannot be null");
+
+    byte[] fanOutBytes = fanOutStore.get(toFanOutStoreKey(taskName));
+    if (ArrayUtils.isEmpty(fanOutBytes)) {
+      return ImmutableMap.of();
+    }
+    StartpointFanOutPerTask startpointFanOutPerTask = objectMapper.readValue(fanOutBytes, StartpointFanOutPerTask.class);
+    return ImmutableMap.copyOf(startpointFanOutPerTask.getFanOuts());
   }
 
   /**
-   * Relinquish resources held by the underlying {@link MetadataStore}
+   * Deletes the fanned out {@link Startpoint} for the given {@link TaskName}
+   * @param taskName to delete the fan out Startpoints for
    */
-  public void stop() {
-    stopped = true;
-    // Metadata store lifecycle is managed outside of the StartpointManager, so not closing it.
+  public void removeFanOutForTask(TaskName taskName) {
+    Preconditions.checkState(!stopped, "Underlying metadata store not available");
+    Preconditions.checkNotNull(taskName, "TaskName cannot be null");
+
+    fanOutStore.delete(toFanOutStoreKey(taskName));
   }
 
   @VisibleForTesting
   MetadataStore getMetadataStore() {
     return metadataStore;
   }
 
-  private static String toStoreKey(SystemStreamPartition ssp, TaskName taskName) {
-    return new String(new JsonSerdeV2<>().toBytes(new StartpointKey(ssp, taskName)));
+  @VisibleForTesting
+  MetadataStore getReadWriteStore() {
+    return readWriteStore;
+  }
+
+  @VisibleForTesting
+  MetadataStore getFanOutStore() {
+    return fanOutStore;
+  }
+
+  @VisibleForTesting
+  ObjectMapper getObjectMapper() {
+    return objectMapper;
+  }
+
+  private static Startpoint resolveStartpointPrecendence(Startpoint startpoint1, Startpoint startpoint2) {
+    if (startpoint1 != null && startpoint2 != null) {
```

Comment:
Since we've capability to resolve startpoint to offset now, we can even support startpoint with broad-cast stream, by resolving the startpoint to offset and comparing the resolved offsets.  What do you think?

Reply:
I'm a little confused. Isn't the offset resolution done in the offset manager? The resolution here is the resolution between two startpoints, not a startpoint and the resolved offset. Also how would we support startpoint with broadcast here? 
In a follow on PR, I'll be integrating the fan out with the JobCoordinators to use the info in the job model to fan out the startpoints, which should take care of broadcast streams. Once the startpoints are fanned out, the actual offset resolution itself happens in the offset manager.